sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/TracezSpanProcessor.java

```diff
@@ -0,0 +1,203 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.extensions.zpages;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.opentelemetry.sdk.common.export.ConfigBuilder;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import io.opentelemetry.trace.SpanId;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A {@link SpanProcessor} implementation for the traceZ zPage.
+ *
+ * <p>Configuration options for {@link TracezSpanProcessor} can be read from system properties,
+ * environment variables, or {@link java.util.Properties} objects.
+ *
+ * <p>For system properties and {@link java.util.Properties} objects, {@link TracezSpanProcessor}
+ * will look for the following names:
+ *
+ * <ul>
+ *   <li>{@code otel.zpages.export.sampled}: sets whether only sampled spans should be exported.
+ * </ul>
+ *
+ * <p>For environment variables, {@link TracezSpanProcessor} will look for the following names:
+ *
+ * <ul>
+ *   <li>{@code OTEL_ZPAGES_EXPORT_SAMPLED}: sets whether only sampled spans should be exported.
+ * </ul>
+ */
+@ThreadSafe
+final class TracezSpanProcessor implements SpanProcessor {
+  private final ConcurrentMap<SpanId, ReadableSpan> runningSpanCache;
+  private final ConcurrentMap<String, TracezSpanBuckets> completedSpanCache;
+  private final boolean sampled;
+
+  /**
+   * Constructor for {@link TracezSpanProcessor}.
+   *
+   * @param sampled report only sampled spans.
+   */
+  public TracezSpanProcessor(boolean sampled) {
+    runningSpanCache = new ConcurrentHashMap<>();
+    completedSpanCache = new ConcurrentHashMap<>();
+    this.sampled = sampled;
+  }
+
+  @Override
+  public void onStart(ReadableSpan span) {
+    runningSpanCache.putIfAbsent(span.getSpanContext().getSpanId(), span);
+  }
+
+  @Override
+  public boolean isStartRequired() {
+    return true;
+  }
+
+  @Override
+  public void onEnd(ReadableSpan span) {
+    runningSpanCache.remove(span.getSpanContext().getSpanId());
+    if (!sampled || span.getSpanContext().getTraceFlags().isSampled()) {
+      completedSpanCache.putIfAbsent(span.getName(), new TracezSpanBuckets());
+      synchronized (this) {
+        completedSpanCache.get(span.getName()).addToBucket(span);
```

Comment:
I don't think we need all this synchronization if we're using a ConcurrentMap, do we?

Reply:
The `addToBucket` function adds to an EvictingQueue, which doesn't seem to be thread-safe.

Reply:
Shouldn't you make the TracezSpanBuckets handle it's own thread-safety, then?

Also, it is generally a bad practice to synchronize on `this` as it introduces the possibility that some other class will synchronize on the same instance, and possibly introduce a deadlock. Any manual synchronziation blocks must use a private final instance variable to lock on.

Reply:
Changed the synchronization targets and made TracezSpanBuckets thread-safe!